lib/segment/src/vector_storage/multi_dense/appendable_mmap_multi_dense_vector_storage.rs

```diff
@@ -0,0 +1,271 @@
+use std::borrow::Cow;
+use std::fs::create_dir_all;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+use bitvec::prelude::BitSlice;
+use common::types::PointOffsetType;
+
+use crate::common::operation_error::{check_process_stopped, OperationResult};
+use crate::common::Flusher;
+use crate::data_types::named_vectors::CowVector;
+use crate::data_types::primitive::PrimitiveVectorElement;
+use crate::data_types::vectors::{
+    MultiDenseVector, TypedMultiDenseVector, TypedMultiDenseVectorRef, VectorElementType, VectorRef,
+};
+use crate::types::{Distance, MultiVectorConfig, VectorStorageDatatype};
+use crate::vector_storage::chunked_mmap_vectors::ChunkedMmapVectors;
+use crate::vector_storage::dense::dynamic_mmap_flags::DynamicMmapFlags;
+use crate::vector_storage::{MultiVectorStorage, VectorStorage, VectorStorageEnum};
+
+const VECTORS_DIR_PATH: &str = "vectors";
+const OFFSETS_DIR_PATH: &str = "offsets";
+const DELETED_DIR_PATH: &str = "deleted";
+
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+struct MultivectorMmapOffset {
+    offset: PointOffsetType,
+    count: PointOffsetType,
+    reserved: PointOffsetType,
+}
+
+pub struct AppendableMmapMultiDenseVectorStorage<T: PrimitiveVectorElement + 'static> {
+    vectors: ChunkedMmapVectors<T>,
+    offsets: ChunkedMmapVectors<MultivectorMmapOffset>,
+    deleted: DynamicMmapFlags,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    deleted_count: usize,
+}
+
+pub fn open_appendable_memmap_multi_vector_storage(
+    path: &Path,
+    dim: usize,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    stopped: &AtomicBool,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    let storage = open_appendable_memmap_multi_vector_storage_impl::<VectorElementType>(
+        path,
+        dim,
+        distance,
+        multi_vector_config,
+        stopped,
+    )?;
+
+    Ok(Arc::new(AtomicRefCell::new(
+        VectorStorageEnum::MultiDenseAppendableMemmap(Box::new(storage)),
+    )))
+}
+
+pub fn open_appendable_memmap_multi_vector_storage_impl<T: PrimitiveVectorElement>(
+    path: &Path,
+    dim: usize,
+    distance: Distance,
+    multi_vector_config: MultiVectorConfig,
+    stopped: &AtomicBool,
+) -> OperationResult<AppendableMmapMultiDenseVectorStorage<T>> {
+    create_dir_all(path)?;
+
+    let vectors_path = path.join(VECTORS_DIR_PATH);
+    let offsets_path = path.join(OFFSETS_DIR_PATH);
+    let deleted_path = path.join(DELETED_DIR_PATH);
+
+    let vectors = ChunkedMmapVectors::open(&vectors_path, dim)?;
+    let offsets = ChunkedMmapVectors::open(&offsets_path, 1)?;
```

Comment:
Ugly moment. I want to reuse extendable mmap for `offsets`. But we have only 1 offset for each multivector. So, I use then `ChunkedMmapVectors` too but with dimension `1`

Reply:
Do you think using `ChunkedMmapVectors` with a dimension of 1 is actually a hack?
What would be an alternative? Building a simpler appendable mmap abstraction without any chunking?

Reply:
```
Building a simpler appendable mmap abstraction without any chunking?
```
Yep, it's an alternative

Reply:
I must admit I don't really like this approach, but I don't have a better suggestion.

`ChunkedMmapVectors` could probably benefit from another round of refactoring to make it a bit more generic (and not vector specific according to it's name). That would probably make it better. It's not something to worry about in this PR though.